File: inflearn_basic_DB/board_dsl/src/main/java/com/dooho/board/api/comment/CommentService.java
package com.dooho.board.api.comment;

import com.dooho.board.api.comment.CommentDto;
import com.dooho.board.api.ResponseDto;
import com.dooho.board.api.comment.PatchCommentDto;
import com.dooho.board.api.comment.PatchCommentResponseDto;
import com.dooho.board.api.board.BoardEntity;
import com.dooho.board.api.comment.CommentEntity;
import com.dooho.board.api.board.BoardRepository;
import com.dooho.board.api.comment.CommentRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Service
public class CommentService {

    private final CommentRepository commentRepository;
    private final BoardRepository boardRepository;

    @Autowired
    public CommentService(CommentRepository commentRepository, BoardRepository boardRepository) {
        this.commentRepository = commentRepository;
        this.boardRepository = boardRepository;
    }


    public ResponseDto<?> register(CommentDto dto){

        CommentEntity commentEntity = new CommentEntity(dto);

        try{
            commentRepository.save(commentEntity);
        }catch (Exception e){
            e.printStackTrace();
            return ResponseDto.setFailed("DataBase Error!");
        }


        return ResponseDto.setSuccess("Success",null);

    }

    public ResponseDto<List<CommentEntity>> getComment(Integer boardNumber){
        List<CommentEntity> commentList = new ArrayList<>();

        BoardEntity boardEntity = boardRepository.findById(boardNumber).orElse(null);
        Integer commentCount = 0;

        try{
            commentList = commentRepository.getComment(boardNumber);
            commentCount = commentRepository.countByBoardNumber(boardNumber);

            boardEntity.setBoardCommentCount(commentCount);
            boardRepository.save(boardEntity);

        }catch (Exception e){
            e.printStackTrace();
            return ResponseDto.setFailed("DataBase Error!");
        }

        return ResponseDto.setSuccess("Success",commentList);


    }

    public ResponseDto<PatchCommentResponseDto> editComment(Integer boardNumber, Integer commentId, PatchCommentDto dto) {
        CommentEntity comment = null;
        String commentContent = dto.getCommentContent();
        LocalDate commentWriteDate = dto.getCommentWriteDate();

        try{
            comment = commentRepository.findById(commentId).orElse(null);
            comment.setCommentContent(commentContent);
            comment.setCommentWriteDate(commentWriteDate);

            commentRepository.save(comment);
        }catch (Exception e){
            e.printStackTrace();
            return ResponseDto.setFailed("DataBase Error!");
        }

        PatchCommentResponseDto patchCommentResponseDto = new PatchCommentResponseDto(comment);

        return ResponseDto.setSuccess("Success!",patchCommentResponseDto);
    }



    public ResponseDto<?> deleteComment(Integer boardNumber, Integer commentId) {
        BoardEntity boardEntity = boardRepository.findById(boardNumber).orElse(null);
        try{
            commentRepository.deleteByCommentId(commentId);
            Integer temp = boardEntity.getBoardCommentCount();
            boardEntity.setBoardCommentCount(temp - 1);
            boardRepository.save(boardEntity);
        }catch (Exception e){
            e.printStackTrace();
            return ResponseDto.setFailed("DataBase Error!");
        }
        return ResponseDto.setSuccess("Success",null);
    }



}

File: board_spring/src/main/java/com/dooho/board/api/comment/CommentService.kt
package com.dooho.board.api.comment

import com.dooho.board.api.ResponseDto
import com.dooho.board.api.ResponseDto.Companion.setFailed
import com.dooho.board.api.ResponseDto.Companion.setSuccess
import com.dooho.board.api.board.BoardRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.time.LocalDate

@Service
class CommentService @Autowired constructor(
    private val commentRepository: CommentRepository,
    private val boardRepository: BoardRepository
) {
    fun register(dto: CommentDto?): ResponseDto<CommentEntity?> {
        val commentEntity = CommentEntity(dto)
        try {
            commentRepository.save(commentEntity)
        } catch (e: Exception) {
            e.printStackTrace()
            return setFailed("DataBase Error!")
        }
        return setSuccess("Success", commentEntity)
    }

    fun getComment(boardNumber: Int?): ResponseDto<List<CommentEntity?>?> {
        var commentList: List<CommentEntity?>? = ArrayList()
        val boardEntity = boardRepository.findById(boardNumber).orElse(null)!!
        var commentCount: Int? = 0
        try {
            commentList = commentRepository.findByBoardNumberOrderByCommentWriteDate(boardNumber)
            commentCount = commentRepository.countByBoardNumber(boardNumber)
            if (commentCount != null) {
                boardEntity.boardCommentCount = commentCount
            }
            boardRepository.save(boardEntity)
        } catch (e: Exception) {
            e.printStackTrace()
            return setFailed("DataBase Error!")
        }
        return setSuccess("Success", commentList)
    }

    fun editComment(boardNumber: Int?, commentId: Int?, dto: PatchCommentDto?): ResponseDto<PatchCommentResponseDto?> {
        var comment: CommentEntity? = null
        val commentContent: String? = dto?.commentContent
        val commentWriteDate: LocalDate? = dto?.commentWriteDate
        try {
            comment = commentRepository.findById(commentId).orElse(null)
            comment!!.commentContent = commentContent
            comment.commentWriteDate = commentWriteDate
            commentRepository.save(comment)
        } catch (e: Exception) {
            e.printStackTrace()
            return setFailed("DataBase Error!")
        }
        val patchCommentResponseDto = PatchCommentResponseDto(comment)
        return setSuccess("Success!", patchCommentResponseDto)
    }

    fun deleteComment(boardNumber: Int?, commentId: Int?): ResponseDto<*> {
        val boardEntity = boardRepository.findById(boardNumber).orElse(null)!!
        try {
            commentRepository.deleteByCommentId(commentId)
            val temp = boardEntity.boardCommentCount
            boardEntity.boardCommentCount = temp?.minus(1)
            boardRepository.save(boardEntity)
        } catch (e: Exception) {
            e.printStackTrace()
            return setFailed<Any>("DataBase Error!")
        }
        return setSuccess<Any?>("Success", null)
    }
}